azurerm/internal/services/datashare/resource_arm_data_share.go

```diff
@@ -0,0 +1,306 @@
+						"name": {
+							Type:         schema.TypeString,
+							Required:     true,
+							ValidateFunc: validate.DataShareSyncName(),
+						},
+
+						"recurrence": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: validation.StringInSlice([]string{
+								string(datashare.Day),
+								string(datashare.Hour),
+							}, false),
+						},
+
+						"start_time": {
+							Type:             schema.TypeString,
+							Required:         true,
+							ValidateFunc:     validation.IsRFC3339Time,
+							DiffSuppressFunc: suppress.RFC3339Time,
+						},
+					},
+				},
+			},
+
+			"terms": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+		},
+	}
+}
+func resourceArmDataShareCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).DataShare.SharesClient
+	syncClient := meta.(*clients.Client).DataShare.SynchronizationClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	name := d.Get("name").(string)
+	accountId, err := parse.DataShareAccountID(d.Get("account_id").(string))
+	if err != nil {
+		return err
+	}
+
+	if d.IsNewResource() {
+		existing, err := client.Get(ctx, accountId.ResourceGroup, accountId.Name, name)
+		if err != nil {
+			if !utils.ResponseWasNotFound(existing.Response) {
+				return fmt.Errorf("checking for present of existing DataShare %q (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+			}
+		}
+		if existing.ID != nil && *existing.ID != "" {
+			return tf.ImportAsExistsError("azurerm_data_share", *existing.ID)
+		}
+	}
+
+	share := datashare.Share{
+		ShareProperties: &datashare.ShareProperties{
+			ShareKind:   datashare.ShareKind(d.Get("share_kind").(string)),
+			Description: utils.String(d.Get("description").(string)),
+			Terms:       utils.String(d.Get("terms").(string)),
+		},
+	}
+
+	if _, err := client.Create(ctx, accountId.ResourceGroup, accountId.Name, name, share); err != nil {
+		return fmt.Errorf("creating DataShare %q (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+	}
+
+	resp, err := client.Get(ctx, accountId.ResourceGroup, accountId.Name, name)
+	if err != nil {
+		return fmt.Errorf("retrieving DataShare %q (Resource Group %q / accountName %q): %+v", name, accountId.ResourceGroup, accountId.Name, err)
+	}
+
+	if resp.ID == nil || *resp.ID == "" {
+		return fmt.Errorf("reading DataShare %q (Resource Group %q / accountName %q): ID is empty", name, accountId.ResourceGroup, accountId.Name)
+	}
+
+	d.SetId(*resp.ID)
+
+	if d.HasChange("snapshot_schedule") {
+		// only one dependent sync setting is allowed in one data share
+		o, _ := d.GetChange("snapshot_schedule")
+		if origins := o.([]interface{}); len(origins) > 0 {
+			syncFuture, err := syncClient.Delete(ctx, accountId.ResourceGroup, accountId.Name, name, origins[0].(map[string]interface{})["name"].(string))
```

Comment:
can we check this statement better`origins[0].(map[string]interface{})["name"].(string)` - ensuring no nils or empty arrays